File: 038/cpp/main.cpp
class Solution {
public:
    vector<string> g1(string a) {
        int i = 0;
        vector<string> r;
        for (int j = 1; j < a.size(); j++) {
            if (a[j] != a[i]) {
                r.push_back(a.substr(i, j - i));
                i = j;
            }
        }
        r.push_back(a.substr(i));
        return r;
    }

    string g2(string a) {
        vector<string> r = g1(a);
        string s;
        for (string &c: r) {
            s += std::to_string(c.size()) + c[0];
        }
        return s;
    }

    string countAndSay(int n) {
        string m = "1";
        for (int i = 1; i < n; i++) {
            m = g2(m);
        }
        return m;
    }
};

File: 038/rust/main.rs
impl Solution {
    fn g1(s: String) -> Vec<String> {
        let mut i = 0;
        let mut r = Vec::new();
        let b = s.as_bytes();
        for j in 0..b.len() {
            if b[j] != b[i] {
                r.push((&s[i..j]).to_string());
                i = j;
            }
        }
        r.push((&s[i..b.len()]).to_string());
        return r;
    }

    fn g2(s: String) -> String {
        let r = Solution::g1(s);
        let mut res = String::new();
        for c in r {
            res += &format!("{}{}", c.len(), c.chars().nth(0).unwrap());
        }
        return res;
    }

    pub fn count_and_say(n: i32) -> String {
        let mut m = "1".to_string();
        for i in 1..n {
            m = Solution::g2(m);
        }
        return m;
    }
}
